gnxi_tester/config/history.go

```diff
@@ -6,28 +21,59 @@ import (
 	"github.com/spf13/viper"
 )
 
-const defaultPort = ":9339"
+// Device stores connection details of a target.
+type Device struct {
+	Address string `mapstructure:"address"`
+	Ca      string `mapstructure:"ca"`
+	CaKey   string `mapstructure:"cakey"`
+}
 
-// SetTarget adds any new target to the target history.
-func SetTarget(targetName, targetAddress string) error {
+// SetTarget adds any new target to the list of known targets.
+func SetTarget(targetName, targetAddress, ca, caKey string) error {
+	if err := prepareTarget(targetName, targetAddress, ca, caKey); err != nil {
+		return err
+	}
+	if err := viper.WriteConfig(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// prepareTarget parses provided details and creates or modifies target entry.
+func prepareTarget(targetName, targetAddress, ca, caKey string) error {
 	devices := GetDevices()
+	if devices == nil {
+		devices = map[string]Device{}
+	}
 	if targetName == "" {
 		if len(devices) > 0 {
 			return nil
 		}
 		return errors.New("No targets in history and no target specified")
 	}
-	if devices[targetName] == "" {
-		if targetAddress == "" {
-			devices[targetName] = defaultPort
-		} else {
-			devices[targetName] = targetAddress
+	if _, exists := devices[targetName]; !exists {
+		if targetAddress == "" || ca == "" || caKey == "" {
```

Comment:
Maybe we should consider storing paths as absolute only. You could use `path.IsAbs` to ensure its absolute and `path.Join` if it's not

Reply:
Sounds good to me!